WalletWasabi/Wallets/BlockDownloadStats.cs

```diff
@@ -0,0 +1,153 @@
+			var nodeDlsHistory = History.Where(x => x.From.Equals(node.RemoteSocketAddress));
+			var nodeDlsHistoryCount = nodeDlsHistory.Count();
+			
+			// KeepNodeScore components, the higher the score the higher probability to keep the node
+			var connectedNodesScore = connectedNodes >= Const.ConnectedNodesGoal
+				? (1 - Math.Min(1, (connectedNodes - Const.ConnectedNodesGoal) * Const.TooMuchNodesScoreMultiplier)) * 100 // Low score when too much nodes
+				: (((double)Const.ConnectedNodesGoal - (connectedNodes - 1)) / Const.ConnectedNodesGoal) * 100; // High score when low number of nodes
+			
+			// Low score if high frequency of dl from this node
+			var nodeFrequencyScore = (1 - Math.Min(1, Math.Pow(nodeDlsSampleCount, 2) / Sample.Count())) * 100;
+
+			// Low score if node tends to timeout
+			var nodeHistoryTimeoutsScore = Math.Max(0.15, 1 - Math.Min(1, nodeDlsHistory.Count(x => !x.Success) * Const.NodeHistoryTimeoutsMultiplier / nodeDlsHistoryCount)) * 100;
+			if (ConsecutiveFails > 1)
+			{
+				// Lot of consecutive fails probably means temporary bandwidth/tor issue, avoid disconnecting.
+				nodeHistoryTimeoutsScore = Math.Min(80, nodeHistoryTimeoutsScore * ConsecutiveFails);
+			}
+			
+			var keepNodeScore = Math.Round(Const.NodeFrequencyWeight * nodeFrequencyScore 
+			                               + Const.ConnectedNodesWeight * connectedNodesScore 
+			                               + Const.NodeHistoryTimeoutsWeight * nodeHistoryTimeoutsScore);
+
+			var result = keepNodeScore <= InsecureRandom.Instance.GetInt(0, 101);
+			var blockDlResultLog = Sample.Last().Success ? "Block dl SUCCESS" : "Block dl FAILED";
+			Logger.LogDebug(result
+				? $"{blockDlResultLog} - Node: Disconnect - Score: {keepNodeScore} - Nb Nodes remaining: {connectedNodes - 1}"
+				: $"{blockDlResultLog} - Node: Keep - Score: {keepNodeScore} - Nb Nodes: {connectedNodes}");
+			return result;
+		}
+	}
+	
+	private class BlockDl
+	{
+		public BlockDl(IPAddress from, double msElapsed, bool success)
+		{
+			From = from;
+			MsElapsed = msElapsed;
+			Success = success;
+		}
+		
+		public IPAddress From { get; }
+		public double MsElapsed { get; }
+		public bool Success { get; }
+	}
+	
+	private static class Const 
+	{
+		// Sample parameters
+		public static int SampleSize => 10;
+		public static int SampleMinSuccesses => 2;
+		
+		// Node timeout strategy parameters
+		public static int BlockDlFailedPenalty => 500;
+		public static double BlockDlFailedMultiplier => 1.5;
+		public static double StandardDerivationWeight => 2;
+		public static int MinimumSpread => 2000;
+		public static int TimeoutDefaultValue => 5000;
+		public static int TimeoutMaxValue => 256000;
+		
+		// Node disconnect strategy parameters
+		public static int ConnectedNodesGoal => 4;
+		public static double TooMuchNodesScoreMultiplier => 0.2;
+		public static double ConnectedNodesWeight => 0.25;
+		
+		public static double NodeFrequencyWeight => 0.1;
```

Comment:
It is hard to distinguish which value is used where. I would the parameters at the begging of the method where are used. publish static makes sense only if these are required from outside.

Reply:
I think this is why you have trouble understanding the algo, because its pretty simple. I will move everything where its needed